internal/compare/compare.go

```diff
@@ -0,0 +1,140 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package compare
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"go.uber.org/thriftrw/compile"
+)
+
+// Diagnostic is a message associated with an error and a file name.
+type Diagnostic struct {
+	File    string // File where error was discovered
+	Message string // Message contains error message.
+}
+
+func (d *Diagnostic) String() string {
+	return fmt.Sprintf("%s:%s", d.File, d.Message)
+}
+
+// Pass provides all reported errors.
+type Pass struct {
+	lints []Diagnostic
+}
+
+// Report reports an error.
+func (p *Pass) Report(d Diagnostic) {
+	p.lints = append(p.lints, d)
+}
+
+// Lints returns all errors.
+func (p *Pass) Lints() []Diagnostic {
+	return p.lints
+}
+
+func (p *Pass) String() string {
+	var b strings.Builder
+	for _, l := range p.lints {
+		_, _ = fmt.Fprintf(&b, "%s\n", l.String())
+	}
+
+	return b.String()
+}
+
+// CompareModules looks for removed methods and added required fields.
+func (p *Pass) CompareModules(from, to *compile.Module) {
+	for name, fromService := range from.Services {
+		p.service(fromService, to.Services[name])
+	}
+
+	for n, fromType := range from.Types {
+		p.typ(fromType, to.Types[n], filepath.Base(from.ThriftPath))
+	}
+}
+
+func (p *Pass) typ(from, to compile.TypeSpec, file string) {
+	if f, ok := from.(*compile.StructSpec); ok {
+		t, ok := to.(*compile.StructSpec)
+		if !ok {
+			// This is a new Type, which is backwards compatible.
+			return
+		}
+		p.structSpecs(f, t, file)
+	}
+}
+
+// StructSpecs compares two structs defined in a Thrift file.
+func (p *Pass) structSpecs(from, to *compile.StructSpec, file string) {
+	fields := make(map[int16]*compile.FieldSpec, len(from.Fields))
+	// Assume that these two should be compared.
+	for _, f := range from.Fields {
+		// Capture state of all fields here.
+		fields[f.ID] = f
+	}
+
+	for _, toField := range to.Fields {
+		if fromField, ok := fields[toField.ID]; ok {
+			fromRequired := fromField.Required
+			toRequired := toField.Required
+			if !fromRequired && toRequired {
+				p.Report(Diagnostic{
+					File: file,
+					Message: fmt.Sprintf(
+						"changing an optional field %s in %s to required is not backwards compatible",
+						toField.ThriftName(), to.ThriftName()),
```

Comment:
I would use `%q` to quote the field names here and elsewhere.

```suggestion
						"changing an optional field %q in %q to required is not backwards compatible",
						toField.ThriftName(), to.ThriftName()),
```